src/random-sleep.1
.\" ==========================================================================
.\"         ____            _                     _____           _
.\"        / ___| _   _ ___| |_ ___ _ __ ___     |_   _|__   ___ | |___
.\"        \___ \| | | / __| __/ _ \ '_ ` _ \ _____| |/ _ \ / _ \| / __|
.\"         ___) | |_| \__ \ ||  __/ | | | | |_____| | (_) | (_) | \__ \
.\"        |____/ \__, |___/\__\___|_| |_| |_|     |_|\___/ \___/|_|___/
.\"               |___/
.\"                             --- System-Tools ---
.\"                  https://www.nntb.no/~dreibh/system-tools/
.\" ==========================================================================
.\"
.\" Random-Sleep
.\" Copyright (C) 2024-2025 by Thomas Dreibholz
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" Contact: thomas.dreibholz@gmail.com
.\"
.\" ###### Setup ############################################################
.Dd May 11, 2025
.Dt random-sleep 1
.Os random-sleep
.\" ###### Name #############################################################
.Sh NAME
.Nm random-sleep
.Nd waits for a randomly selected time
.\" ###### Synopsis #########################################################
.\" Manpage syntax help:
.\" https://forums.freebsd.org/threads/howto-create-a-manpage-from-scratch.13200/
.Sh SYNOPSIS
.Nm random-sleep
min_delay
max_delay
.Op Fl q | Fl Fl quiet
.Op Fl w | Fl Fl verbose
.Nm random-sleep
.Op Fl h | Fl Fl help
.Nm random-sleep
.Op Fl v | Fl Fl version
.\" ###### Description ######################################################
.Sh DESCRIPTION
.Nm random-sleep
waits for a randomly selected time. The sleep time is chosen from the interval
[min_delay, max_delay] given in seconds. Fractional seconds are supported.
.Pp
.\" ###### Arguments ########################################################
.Sh ARGUMENTS
The following options are available:
.Bl -tag -width indent
.It Fl q | Fl Fl quiet
Quiet mode: no output is made.
.It Fl w | Fl Fl verbose
Verbose mode: print selected delay and wake-up.
.It Fl h | Fl Fl help
Prints command-line parameters.
.It Fl v | Fl Fl version
Prints program version.
.El
.\" ###### Examples #########################################################
.Sh EXAMPLES
.Bl -tag -width indent
.It random-sleep 0 3600
.It random-sleep 0 300 --verbose
.It random-sleep 60.0 120.5 -w
.It random-sleep 0.10 0.50 --verbose
.It random-sleep --version
.It random-sleep --help
.El


ChangeLog
commit 69294977b70873d0d5e7c78447da495134900a69
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon May 12 22:24:25 2025 +0200

    Updated CI configuration

commit 48b58d6e44d1597bba157b089fd4d88c7ef15d63
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon May 12 22:24:22 2025 +0200

    Updated CI scripts

commit 4f90161dd8b4a76f423a050987ea45904b593ddf
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon May 12 22:24:19 2025 +0200

    Updated build scripts

commit 0d347201f4ffe1173f7884674c51d76fbb6ff3db
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sun May 11 21:02:12 2025 +0200

    Improved manpage. Also fixed too-long lines with hyphenation suggestions \% => no need for Lintian override any more.

commit fa24c68be02a7dfca8812eb4cccafaedb70c7d23
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat May 10 17:07:06 2025 +0200

    Removed debug output.

commit 712ad3f568e7a633aade4dd0650d5d8f063f65c1
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat May 10 11:42:31 2025 +0200

    Added missing Text-Block options to shell completion.

commit 62da1ead12bb5af8efd2a3975ce2d15ba21baadd
Merge: 0103c17 3e5c731
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat May 10 11:41:32 2025 +0200

    Merge branch 'dreibh/text-block'

commit 3e5c731430793a9a53a2d12b779ffb7b552b7fc5
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat May 10 11:19:03 2025 +0200

    Updated test script.

commit b43e614ea08c702d018583893cf2a904bd616f7a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat May 10 11:15:50 2025 +0200

    Bugfix: Separator must have a separator string of width >= 1, in order to not cause an infinite loop.

commit 3314cc5aedc15ba15c6e3b94fb4ce3a56015c6e1
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat May 10 10:34:44 2025 +0200

    Bugfix: Fixed possible memory leaks found by static analysis.

commit a445cdb71f99c6776f3242ee1aba4f865b85df54
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat May 10 10:17:39 2025 +0200

    Bugfix: Fixed potential error cases found by static analysis.

commit bad03668230a56adbf2e081f291729a9304ceebe
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat May 10 10:05:15 2025 +0200

    Improved in-place and keep-broken options handling.

commit 828d25bbd8b2d4fdd1f7ef64ded53b4ba63c4ffa
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri May 9 08:54:51 2025 +0200

    Some work on update options.

commit 0103c17d1f97ed6d9f1b0e9c9d2dd4bd9272671b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 8 22:45:42 2025 +0200

    Minor fix.

commit c122646e6fe56ba0910c5b832d97864b734a0f05
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 8 22:44:40 2025 +0200

    Some message clean-ups.

commit 604701627678832cc43bd83f4683e1a458b8ffed
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 8 22:25:08 2025 +0200

    Added missing insert/replace from standard input.

commit f106ac799ce228e5157fcd3e2fab206b52aeff27
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sun Apr 27 20:19:29 2025 +0200

    Updated FreeBSD port.

commit 8ff347c38ce733a5455ff54ef75da871deacc419
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sun Apr 27 14:15:09 2025 +0200

    New release td-system-tools-2.1.0.

commit aefe815f79164563e32917e464225c768a1523d8
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sun Apr 27 12:09:31 2025 +0200

    Minor documentation update.

commit bf30a3154ffee25f631ca595327da3cfae0eb885
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sun Apr 27 11:57:08 2025 +0200

    Improved documentation.

commit 79b9310235cb290d2d4164c2aeb74637c213dcf2
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sun Apr 27 11:10:16 2025 +0200

    Added test script.

commit b430b49339e4585e2010b3f713343162891c3324
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sun Apr 27 10:29:20 2025 +0200

    Improved documentation.

commit bda30e9c8c423d7ca9a8e6c57e5fd480454e2b11
Merge: 7dd7ab2 7a8d26b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Apr 26 15:03:54 2025 +0200

    Merge branch 'master' into dreibh/text-block

commit 7a8d26b04042b5b47dd70a686babc510d1f008c8
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Apr 26 14:58:24 2025 +0200

    New release td-system-tools-2.0.10.

commit 7dd7ab2ed59bccea6370bdcb6915d1238a9b65d1
Merge: baaca99 d8719e1
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Apr 26 14:48:22 2025 +0200

    Merge branch 'master' into dreibh/text-block

commit d8719e1bad49eac1982bb5e2ed3f183a89e028ea
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Apr 26 12:47:39 2025 +0200

    Manpage fix.

commit baaca996c27fc1655470327f0af9a5a2fef4f620
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Apr 26 12:36:44 2025 +0200

    Header updates.

commit 0225430b53775643225e8daa31d78604e4b26133
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Apr 26 12:36:26 2025 +0200

    Improved input file and options handling in CMakeLists.txt.

commit d6fb77d9a4593a3a637eb1013c4548c00fdb5fa0
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Apr 26 11:44:53 2025 +0200

    Header updates.

commit 49741d2495b9fa1feb0f33eaf6c4585c92970db1
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Apr 26 11:39:12 2025 +0200

    CMakeLists.txt improvement: Using --no-wrap for xgettext and msgmerge.

commit 304f26c55a4ef71e22f513dba9cac53ae8d8b58e
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Apr 26 11:39:05 2025 +0200

    Updated translations.

commit e8c4c116f8ad574ea603a2c276bb9763ed281d7f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Apr 26 11:38:46 2025 +0200

    Improved output messages.

commit b3b82e364f4572aa0689b4e4a2ba6f0026929b73
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Apr 26 10:44:48 2025 +0200

    Some improvements.

commit 1c717c64dd5641171e8ab925e70ae62bf8a81e9b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Apr 25 17:50:16 2025 +0200

    Another update.


src/random-sleep.c
// ==========================================================================
//         ____            _                     _____           _
//        / ___| _   _ ___| |_ ___ _ __ ___     |_   _|__   ___ | |___
//        \___ \| | | / __| __/ _ \ '_ ` _ \ _____| |/ _ \ / _ \| / __|
//         ___) | |_| \__ \ ||  __/ | | | | |_____| | (_) | (_) | \__ \
//        |____/ \__, |___/\__\___|_| |_| |_|     |_|\___/ \___/|_|___/
//               |___/
//                             --- System-Tools ---
//                  https://www.nntb.no/~dreibh/system-tools/
// ==========================================================================
//
// Random-Sleep
// Copyright (C) 2024-2025 by Thomas Dreibholz
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// Contact: thomas.dreibholz@gmail.com

#include <getopt.h>
#include <libintl.h>
#include <locale.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include <unistd.h>
#ifndef nullptr
#define nullptr NULL
#endif

#include "package-version.h"


// ###### Generate random number out of [min, max] ##########################
static double runiform(const double min, const  double max)
{
   if(min == max) {
      // There is no other choice here:
      return min;
   }
   const double range   = (max - min);
   const double divisor = RAND_MAX / range;
   return min + (rand() / divisor);
}


// ###### Version ###########################################################
[[ noreturn ]] static void version()
{
   printf("random-sleep %s\n", SYSTEMTOOLS_VERSION);
   exit(0);
}


// ###### Usage #############################################################
[[ noreturn ]] static void usage(const char* program, const int exitCode)
{
   fprintf(stderr, "%s %s min_delay max_delay"
           " [-q|--quiet]"
           " [-w|--verbose]"
           " [-h|--help]"
           " [-v|--version]\n",
           gettext("Usage:"), program);
   exit(exitCode);
}



// ###### Main program ######################################################
int main(int argc, char** argv)
{
   double delayMin = -1.0;
   double delayMax = -1.0;
   bool   verboseMode = false;

   // ====== Initialise i18n support ========================================
   if(setlocale(LC_ALL, "") == nullptr) {
      setlocale(LC_ALL, "C.UTF-8");   // "C" should exist on all systems!
   }
   bindtextdomain("random-sleep", nullptr);
   textdomain("random-sleep");

   // ====== Handle arguments ===============================================
   const static struct option long_options[] = {
      { "verbose", no_argument, 0, 'w' },
      { "quiet",   no_argument, 0, 'q' },
      { "help",    no_argument, 0, 'h' },
      { "version", no_argument, 0, 'v' },
      {  nullptr,  0,           0, 0   }
   };

   if(optind + 1 < argc) {
      delayMin = atof(argv[optind + 0]);
      delayMax = atof(argv[optind + 1]);
      optind += 2;
   }

   int option;
   int longIndex;
   while( (option = getopt_long(argc, argv, "wqhv", long_options, &longIndex)) != -1 ) {
      switch(option) {
         case 'w':
            verboseMode = true;
          break;
         case 'q':
            verboseMode = false;
          break;
         case 'v':
            version();
          break;
         case 'h':
            usage(argv[0], 0);
          break;
         default:
            fprintf(stderr, gettext("ERROR: Invalid argument %s!"), argv[optind - 1]);
            fputs("\n", stderr);
            return 1;
          break;
      }
   }
   if(optind < argc) {
      usage(argv[0], 1);
   }
   if( (delayMin < 0.0) || (delayMin > delayMax) ) {
      fputs(gettext("ERROR: Invalid min_delay/max_delay"), stderr);
      fputs("\n", stderr);
      return 1;
   }

   // ====== Initialise random number generator =============================
   struct timeval tv;
   if(gettimeofday(&tv, nullptr) == 0) {
      srand(tv.tv_usec);
   }

   // ====== Random sleep ===================================================
   double delay = runiform(delayMin, delayMax);
   if(delay < 0.0) {
      delay = 0.0;
   }

   if(verboseMode) {
      printf(gettext("Sleeping for %1.6f s ..."), delay);
      puts("");
   }
   usleep((useconds_t)(1000000.0 * delay));
   if(verboseMode) {
      puts(gettext("Woke up!"));
   }

   return 0;
}
